File: apps/api/vfdelivery/models/order.py
import uuid
from datetime import datetime
from enum import Enum
from typing import TYPE_CHECKING, List

from sqlalchemy import ForeignKey, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from vfdelivery.core.database import table_registry

if TYPE_CHECKING:
    from vfdelivery.models.order_item import OrderItem
    from vfdelivery.models.restaurant import Restaurant
    from vfdelivery.models.user import User


class OrderStatus(str, Enum):
    CREATED = 'created'
    ACCEPTED = 'accepted'
    REJECTED = 'rejected'
    DELIVERED = 'delivered'


@table_registry.mapped_as_dataclass
class Order:
    __tablename__ = 'tb_orders'

    id: Mapped[uuid.UUID] = mapped_column(
        init=False,
        primary_key=True,
        default_factory=uuid.uuid4
    )

    customer_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey('tb_users.id'),
        nullable=False
    )

    restaurant_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey('tb_restaurants.id'),
        nullable=False
    )

    total_price: Mapped[float] = mapped_column(
        nullable=False
    )

    status: Mapped[OrderStatus] = mapped_column(
        nullable=False,
        default=OrderStatus.CREATED
    )

    created_at: Mapped[datetime] = mapped_column(
        init=False,
        server_default=func.now()
    )

    updated_at: Mapped[datetime] = mapped_column(
        init=False,
        server_default=func.now(),
        onupdate=func.now()
    )

    customer: Mapped['User'] = relationship(init=False)
    restaurant: Mapped['Restaurant'] = relationship(init=False)
    items: Mapped[List['OrderItem']] = relationship(init=False)

File: apps/web/src/features/customer/types/order.ts
export type CreateOrderItemPayload = {
  product_id: string;
  quantity: number;
};

export type CreateOrderPayload = {
  restaurant_id: string;
  items: CreateOrderItemPayload[];
};

export type OrderStatus = 'created' | 'accepted' | 'rejected' | 'delivered' | 'canceled';

export interface OrderItem {
  id: string;
  product_id: string;
  product_name: string;
  quantity: number;
  unit_price: number;
  subtotal: number;
}

export interface CustomerOrder {
  id: string;
  status: OrderStatus;
  total_price: number;
  created_at: string;
  updated_at: string;
  reviewed: boolean;
  customer: {
    id: string;
    name: string;
  };
  restaurant: {
    id: string;
    name: string;
  };
  items: OrderItem[];
}
